### ISSUE TITLE
We need better commit messages

### ISSUE THREAD

**Issue by kohr-h** [odlgroup/odl#4]
There are a couple of commit messages like "merge" or "bugfix" in the commit history which are actually worse than no message at all. We need to become better at this if we want our Git history to be searchable and useful.

As a very good reference, one can once again at [how NumPy does it](http://docs.scipy.org/doc/numpy-dev/dev/gitwash/development_workflow.html#writing-the-commit-message). I suggest that we follow the same nomenclature, at least in the final versions of pull requests. Intermediate commits on branches may have bad messages, as long as they are cleaned up with `git rebase -i` before the pull request is sent (or at the **very** latest before it is merged into `master`).

Short summary:
- Start each commit message with a capitalized acronym indicating its purpose, e.g. `BUG` for a bugfix (see the table in the link).
- Write a **short** summary (max ~50 characters) on the first line.
  - Use *imperative* style, i.e. it should be possible to append the message (w/o acronym) to "When applied, this commit will ...". In particular, don't write in past tense (`add feature xyz` instead of `added feature xyz`). A very simple reason for this style is that you save on average 2 characters for something else.
    - Be as precise as possible (needs thinking).
    - Don't use markdown.
- Optional: write an extended summary if additional information is needed.
    - Separate it by a blank line from the first line. 
    - Explain why the commit is necessary, e.g. detail the problem which is fixed.
    - You can use markdown here, e.g. include item lists, tables, links, @mentions, ...
    - Restrict the lines to 72 characters in length.
- Refer to tickets (issues) if existing, e.g. "See #2", "Closes #2" (this will actually close the issue) or similar

**Comment by Berzerka** [comment]
This ticket, while relevant, should be part of the documentation. I recommend we close it once #2 is done.

**Comment by adler-j** [comment]
Now that we have a `CONTRIBUTING.md` file as well as an actual documentation. Can we move this to some of those?

**Comment by kohr-h** [comment]
Yes, that's where the information is supposed to be.

**Comment by adler-j** [comment]
Can this be closed now that we have the contributing doc and also gitwash doc?
